### PR TITLE
DeathMatch Rolling Antag Fix

### LINKED ISSUE #81725: It is possible to become a syndicate sleeper agent as a deathmatch player
## Issue Summary
It is possible to become a midround traitor as a deathmatch player

## Round ID: 
[224721](https://scrubby.melonmesa.com/round/224721)

## Reproduction:
Play the deathmatch minigame and get lucky enough to roll midround traitor

### PR DESCRIPTION
**Pull request by Bilbo367**
## About The Pull Request

Gives Deathmatch players the TRAIT_TEMPORARY_BODY which is checked during antag rolls while trimming candidates.

## Why It's Good For The Game

Closes https://github.com/tgstation/tgstation/issues/81725

## Changelog

:cl:
fix: Deathmatch players should no longer be able to roll any midround antagonist.
/:cl:

### CODE REVIEW

**Review by JohnFulpWillard — COMMENTED** [review]
There's something a little deeper I want to address here.
If the player has a body to go back to once the lobby is over, they are given the ``temporary_body`` component, which adds ``TRAIT_MIND_TEMPORARILY_GONE`` to their old body , which trims them out (which doesn't have a mind, so I dont know why lol). ``TRAIT_TEMPORARY_BODY`` is used for the body being being created, but currently it is only used for Bitrunners, and while it is fine to add it here, it doesn't fix it for say, Mafia, Basketball, etc.

The alternative I had in mind is the var ``restrict_ghost_roles``, which currently only excludes you if your assigned role is that of things like Ash Walker, Golem, Hermit, etc. I think it would be better if this simply excluded anyone that isn't an on-station job. This is not really something that should be done on a case-by-case basis, I don't think we ever want non-station roles to roll antagonist, they either are made as one (like Paradox clone, Nukie), or is someone that isn't supposed to be one (drone, Mafia, Oldstation Engineer) because they are meant to be things you do while waiting for the current round to end, not an opportunity to get back into the round and especially not a way to opt into a new antag roll.

**Comment by Bilbo367** [comment]
> , I don't think we ever want non-station roles to roll antagonist, they either are made as one (like Paradox clone, Nukie), or

So if i'm understanding this correctly you want instead check if someone has a station job to get antag instead of removing the jobs that match EXP_TYPE_SPECIAL. This would also accomplish the // TO DO: Replace this with job datum flags instead. and make it so every new station job we add we must add to this list so they can roll midround. Instead of adding every new ghost role.

**Comment by Bilbo367** [comment]
image

**Comment by JohnFulpWillard** [comment]
> So if i'm understanding this correctly you want instead check if someone has a station job to get antag instead of removing the jobs that match EXP_TYPE_SPECIAL. This would also accomplish the // TO DO: Replace this with job datum flags instead. and make it so every new station job we add we must add to this list so they can roll midround. Instead of adding every new ghost role.

What I am saying is that I think a better solution long term would be to make ``restrict_ghost_roles`` check to see if you are a station job, and if you aren't to trim you out.

**Comment by san7890** [comment]
undraft when you've addressed willard's review

image

**Comment by Bilbo367** [comment]
It was a foolish move to merge master that many times.

File: code/controllers/subsystem/dynamic/dynamic_rulesets_midround.dm [hunk 1]
```diff
@@ -72,11 +129,9 @@
 		if (is_banned_from(creature.ckey, list(antag_flag_override || antag_flag, ROLE_SYNDICATE)))
 			trimmed_list.Remove(creature)
 			continue
-
 		if (isnull(creature.mind))
 			continue
-
-		if (restrict_ghost_roles && (creature.mind.assigned_role.title in GLOB.exp_specialmap[EXP_TYPE_SPECIAL])) // Are they playing a ghost role?
+		if (restrict_ghost_roles && !(creature.mind.assigned_role.title in station_jobs)) // Are they not playing a station role?
```
Comment (JohnFulpWillard) [inline code/controllers/subsystem/dynamic/dynamic_rulesets_midround.dm:134]: You could just....
```suggestion
		if (restrict_ghost_roles && !(creature.mind.assigned_role.job_flags & JOB_CREW_MEMBER)) // Are they not playing a station role?
```
And you won't need this massive list

**Pull request merged**